samples/cxx/custom/custom.cpp

```diff
@@ -0,0 +1,197 @@
+        // similarly, the time derivative of the solution vector, *ydot*, is
+        // [dT/dt, Y_1/dt, Y_2/dt, ... Y_K/dt].
+        // the following variables are defined for clear and convenient access
+        // to these vectors:
+        double temperature = y[0];
+        double *massFracs = &y[1];
+        double *dTdt = &ydot[0];
+        double *dYdt = &ydot[1];
+        
+        /* ------------------------- UPDATE GAS STATE ------------------------- */
+        // the state of the ThermoPhase is updated to reflect the current solution
+        // vector, which was calculated by the integrator.
+        gas->setMassFractions_NoNorm(massFracs);
+        gas->setState_TP(temperature, pressure);
+
+        /* ----------------------- GET REQ'D PROPERTIES ----------------------- */
+        double rho = gas->density();
+        double cp = gas->cp_mass();
+        gas->getPartialMolarEnthalpies(&hbar[0]);
+        kinetics->getNetProductionRates(&wdot[0]);
+        
+        /* -------------------------- ENERGY EQUATION ------------------------- */
+        // the rate of change of the system temperature is found using the energy
+        // equation for a closed-system constant pressure ideal gas:
+        //     m*cp*dT/dt = - sum[h(k) * dm(k)/dt]
+        // or equivalently:
+        //     dT/dt = - sum[hbar(k) * dw(k)/dt] / (rho * cp)
+        double hdot_vol = 0;
+        for (int k = 0; k < nSpecies; k++)
+            hdot_vol += hbar[k] * wdot[k];
+        *dTdt = - hdot_vol / (rho * cp);
+
+        /* --------------------- SPECIES CONSERVATION EQS --------------------- */
+        // the rate of change of each species' mass fraction is found using the closed-system
+        // species conservation equation, applied once for each species:
+        //     m*dY(k)/dt = dm(k)/dt
+        // or equivalently:
+        //     dY(k)/dt = dw(k)/dt * MW(k) / rho
+        for (int k = 0; k < nSpecies; k++)
+            dYdt[k] = wdot[k] * gas->molecularWeight(k) / rho;
+    }
+    
+    /**
+     * Number of equations in the ODE system.
+     *   - overrided from FuncEval, called by the integrator during initialization.
+     */
+    size_t neq() {
+        return nEqs;
+    }
+
+    /**
+     * Provide the current values of the state vector, *y*.
+     *   - overrided from FuncEval, called by the integrator during initialization.
+     * @param[out] y solution vector, length neq()
+     */
+    void getState(double* y) {
+        // the solution vector *y* is [T, Y_1, Y_2, ... Y_K], where T is the
+        // system temperature, and Y_k is the mass fraction of species k.
+        y[0] = gas->temperature();
+        gas->getMassFractions(&y[1]);
+    }
+
+private:
+    // private member variables, to be used internally.
+    shared_ptr<ThermoPhase> gas;
+    shared_ptr<Kinetics> kinetics;
+    vector_fp hbar;
+    vector_fp wdot;
+    double pressure;
+    int nSpecies;
+    int nEqs;
+};
```

Comment:
This isn't critical because this is a sample rather than core Cantera code, but we usually prefix member variables with `m_`, which makes it easier to differentiate between members and locals when reading the code in member functions.